File: modules/Scripting/annotations/RmlUi/TransformPrimitives.lua
---@meta

---@class ui.NumericValue
---@field number number
---@field unit ui.Unit
---@overload fun(): ui.NumericValue
---@overload fun(number: number, unit: ui.Unit): ui.NumericValue
ui.NumericValue = {}

ui.Transforms = {}

---@class ui.Transforms.Matrix2D
---@overload fun(values: ui.NumericValue[]): ui.Transforms.Matrix2D
ui.Transforms.Matrix2D = {}

---@class ui.Transforms.Matrix3D
---@overload fun(values: ui.NumericValue[]): ui.Transforms.Matrix3D
ui.Transforms.Matrix3D = {}

---@class ui.Transforms.TranslateX
---@overload fun(values: ui.NumericValue[]): ui.Transforms.TranslateX
---@overload fun(x: number, unit: ui.Unit): ui.Transforms.TranslateX
ui.Transforms.TranslateX = {}

---@class ui.Transforms.TranslateY
---@overload fun(values: ui.NumericValue[]): ui.Transforms.TranslateY
---@overload fun(x: number, unit: ui.Unit): ui.Transforms.TranslateY
ui.Transforms.TranslateY = {}

---@class ui.Transforms.TranslateZ
---@overload fun(values: ui.NumericValue[]): ui.Transforms.TranslateZ
---@overload fun(x: number, unit: ui.Unit): ui.Transforms.TranslateZ
ui.Transforms.TranslateZ = {}

---@class ui.Transforms.Translate2D
---@overload fun(values: ui.NumericValue[]): ui.Transforms.Translate2D
---@overload fun(x: number, y: number, unit: ui.Unit): ui.Transforms.Translate2D
ui.Transforms.Translate2D = {}

---@class ui.Transforms.Translate3D
---@overload fun(values: ui.NumericValue[]): ui.Transforms.Translate3D
---@overload fun(x: number, y: number, z: number, unit: ui.Unit): ui.Transforms.Translate3D
ui.Transforms.Translate3D = {}

---@class ui.Transforms.ScaleX
---@overload fun(values: ui.NumericValue[]): ui.Transforms.ScaleX
---@overload fun(values: number): ui.Transforms.ScaleX
ui.Transforms.ScaleX = {}

---@class ui.Transforms.ScaleY
---@overload fun(values: ui.NumericValue[]): ui.Transforms.ScaleY
---@overload fun(values: number): ui.Transforms.ScaleY
ui.Transforms.ScaleY = {}

---@class ui.Transforms.ScaleZ
---@overload fun(values: ui.NumericValue[]): ui.Transforms.ScaleZ
---@overload fun(values: number): ui.Transforms.ScaleZ
ui.Transforms.ScaleZ = {}

---@class ui.Transforms.Scale2D
---@overload fun(values: ui.NumericValue[]): ui.Transforms.Scale2D
---@overload fun(xy: number): ui.Transforms.Scale2D
---@overload fun(x: number, y: number): ui.Transforms.Scale2D
ui.Transforms.Scale2D = {}

---@class ui.Transforms.Scale3D
---@overload fun(values: ui.NumericValue[]): ui.Transforms.Scale3D
---@overload fun(xyz: number): ui.Transforms.Scale3D
---@overload fun(x: number, y: number, z: number): ui.Transforms.Scale3D
ui.Transforms.Scale3D = {}

---@class ui.Transforms.RotateX
---@overload fun(values: ui.NumericValue[]): ui.Transforms.RotateX
---@overload fun(angle: number, unit: ui.Unit): ui.Transforms.RotateX
ui.Transforms.RotateX = {}

---@class ui.Transforms.RotateY
---@overload fun(values: ui.NumericValue[]): ui.Transforms.RotateY
---@overload fun(angle: number, unit: ui.Unit): ui.Transforms.RotateY
ui.Transforms.RotateY = {}

---@class ui.Transforms.RotateZ
---@overload fun(values: ui.NumericValue[]): ui.Transforms.RotateZ
---@overload fun(angle: number, unit: ui.Unit): ui.Transforms.RotateZ
ui.Transforms.RotateZ = {}

---@class ui.Transforms.Rotate2D
---@overload fun(values: ui.NumericValue[]): ui.Transforms.Rotate2D
---@overload fun(angle: number, unit: ui.Unit): ui.Transforms.Rotate2D
ui.Transforms.Rotate2D = {}

---@class ui.Transforms.Rotate3D
---@overload fun(values: ui.NumericValue[]): ui.Transforms.Rotate3D
---@overload fun(xy: number, y: number, z: number, angle: number, unit: ui.Unit): ui.Transforms.Rotate3D
ui.Transforms.Rotate3D = {}

---@class ui.Transforms.SkewX
---@overload fun(values: ui.NumericValue[]): ui.Transforms.SkewX
---@overload fun(angle: number, unit: ui.Unit): ui.Transforms.SkewX
ui.Transforms.SkewX = {}

---@class ui.Transforms.SkewY
---@overload fun(values: ui.NumericValue[]): ui.Transforms.SkewY
---@overload fun(angle: number, unit: ui.Unit): ui.Transforms.SkewY
ui.Transforms.SkewY = {}

---@class ui.Transforms.Skew2D
---@overload fun(values: ui.NumericValue[]): ui.Transforms.Skew2D
---@overload fun(x: number, y: number, unit: ui.Unit): ui.Transforms.Skew2D
ui.Transforms.Skew2D = {}

---@class ui.Transforms.Perspective
---@overload fun(values: ui.NumericValue[]): ui.Transforms.Perspective
ui.Transforms.Perspective = {}

---@class ui.Transforms.DecomposedMatrix4
ui.Transforms.DecomposedMatrix4 = {}

---@enum ui.TransformPrimitive.Type
ui.TransformPrimitive.Type = {
    MATRIX2D = 0,
    MATRIX3D = 1,
    TRANSLATEX = 2,
    TRANSLATEY = 3,
    TRANSLATEZ = 4,
    TRANSLATE2D = 5,
    TRANSLATE3D = 6,
    SCALEX = 7,
    SCALEY = 8,
    SCALEZ = 9,
    SCALE2D = 10,
    SCALE3D = 11,
    ROTATEX = 12,
    ROTATEY = 13,
    ROTATEZ = 14,
    ROTATE2D = 15,
    ROTATE3D = 16,
    SKEWX = 17,
    SKEWY = 18,
    SKEW2D = 19,
    PERSPECTIVE = 20,
    DECOMPOSEDMATRIX4 = 21,
}

---@class ui.TransformPrimitive
---@field type ui.TransformPrimitive.Type
---@overload fun(p: ui.Transforms.Matrix2D|ui.Transforms.Matrix3D|ui.Transforms.TranslateX|ui.Transforms.TranslateY|ui.Transforms.TranslateZ|ui.Transforms.Translate2D|ui.Transforms.Translate3D|ui.Transforms.ScaleX|ui.Transforms.ScaleY|ui.Transforms.ScaleZ|ui.Transforms.Scale2D|ui.Transforms.Scale3D|ui.Transforms.SkewX|ui.Transforms.SkewY|ui.Transforms.Skew2D|ui.Transforms.Perspective|ui.Transforms.DecomposedMatrix4): ui.TransformPrimitive
ui.Transforms.TransformPrimitive = {}

---@alias ui.PrimitiveList ui.TransformPrimitive[]

---@class ui.Transform
---@overload fun(primitives?: ui.PrimitiveList) ui.Transform
ui.Transform = {}

---Helper function to create a 'transform' Property from the given list of primitives.
---@param primitives ui.PrimitiveList
---@return ui.Property
function ui.Transform.makeProperty(primitives) end

---Remove all Primitives from this Transform.
function ui.Transform:clearPrimitives() end

---Add a Primitive to this Transform
---@param p ui.TransformPrimitive
function ui.Transform:addPrimitive(p) end

---Return the number of Primitives in this Transform.
---@return integer
function ui.Transform:getNumPrimitives() end

---Return the i-th Primitive in this Transform.
---@param i integer
---@return ui.TransformPrimitive
function ui.Transform:getPrimitive(i) end

---@return ui.PrimitiveList
function ui.Transform:getPrimitives() end

File: modules/Scripting/modules/RmlUi/src/TransformPrimitives.cpp
#include "TransformPrimitives.hpp"
#include "RmlUi/Core/Transform.h"
#include "RmlUi/Core/Property.h"

#include "Sol2HelperMacros.hpp"

using namespace Rml;

namespace {

void registerTransformPrimitives(sol::table &lua) {
  // clang-format off
  lua.DEFINE_USERTYPE(NumericValue,
    sol::call_constructor,
    sol::constructors<
      NumericValue(),
      NumericValue(float, Unit)
    >(),

    "number", &NumericValue::number,
    "unit", &NumericValue::unit,

    BIND_TOSTRING(NumericValue)
  );

  using namespace Transforms;

  lua.DEFINE_USERTYPE(Matrix2D,
    sol::call_constructor,
    sol::constructors<Matrix2D(const NumericValue *)>(),
    BIND_TOSTRING(Matrix2D)
  );
  lua.DEFINE_USERTYPE(Matrix3D,
    sol::call_constructor,
    sol::constructors<Matrix3D(const NumericValue *)>(),
    BIND_TOSTRING(Matrix3D)
  );

#define DEFINE_NUMERIC(Type, ...)                                              \
  lua.DEFINE_USERTYPE(Type,                                                        \
    sol::call_constructor,                                                     \
    sol::constructors<Type(const NumericValue *), __VA_ARGS__>(),              \
    BIND_TOSTRING(Type))

  DEFINE_NUMERIC(TranslateX, TranslateX(float, Unit));
  DEFINE_NUMERIC(TranslateY, TranslateY(float, Unit));
  DEFINE_NUMERIC(TranslateZ, TranslateZ(float, Unit));
  DEFINE_NUMERIC(Translate2D, Translate2D(float, float, Unit));
  DEFINE_NUMERIC(Translate3D, Translate3D(float, float, float, Unit));

  DEFINE_NUMERIC(ScaleX, ScaleX(float));
  DEFINE_NUMERIC(ScaleY, ScaleY(float));
  DEFINE_NUMERIC(ScaleZ, ScaleZ(float));
  DEFINE_NUMERIC(Scale2D, Scale2D(float), Scale2D(float, float));
  DEFINE_NUMERIC(Scale3D, Scale3D(float), Scale3D(float, float, float));

  DEFINE_NUMERIC(RotateX, RotateX(float, Unit));
  DEFINE_NUMERIC(RotateY, RotateY(float, Unit));
  DEFINE_NUMERIC(RotateZ, RotateZ(float, Unit));
  DEFINE_NUMERIC(Rotate2D, Rotate2D(float, Unit));
  DEFINE_NUMERIC(Rotate3D, Rotate3D(float, float, float, float, Unit));
  
  DEFINE_NUMERIC(SkewX, SkewX(float, Unit));
  DEFINE_NUMERIC(SkewY, SkewY(float, Unit));
  DEFINE_NUMERIC(Skew2D, Skew2D(float, float, Unit));

#undef DEFINE_NUMERIC

  lua.DEFINE_USERTYPE(Perspective,
    sol::call_constructor,
    sol::constructors<Perspective(const NumericValue *)>(),
    BIND_TOSTRING(Perspective)
  );
  // clang-format on
}

} // namespace

void registerTransform(sol::table &lua) {
  auto m = lua["Transforms"].get_or_create<sol::table>();
  registerTransformPrimitives(m);

  // clang-format off
  lua.DEFINE_USERTYPE(TransformPrimitive,
    sol::call_constructor,
    sol::constructors<
      TransformPrimitive(Transforms::Matrix2D),
      TransformPrimitive(Transforms::Matrix3D),
      TransformPrimitive(Transforms::TranslateX),
      TransformPrimitive(Transforms::TranslateY),
      TransformPrimitive(Transforms::TranslateZ),
      TransformPrimitive(Transforms::Translate2D),
      TransformPrimitive(Transforms::Translate3D),
      TransformPrimitive(Transforms::ScaleX),
      TransformPrimitive(Transforms::ScaleY),
      TransformPrimitive(Transforms::ScaleZ),
      TransformPrimitive(Transforms::Scale2D),
      TransformPrimitive(Transforms::Scale3D),
      TransformPrimitive(Transforms::SkewX),
      TransformPrimitive(Transforms::SkewY),
      TransformPrimitive(Transforms::Skew2D),
      TransformPrimitive(Transforms::Perspective),
      TransformPrimitive(Transforms::DecomposedMatrix4)
    >(),

    _BIND(TransformPrimitive, type),

    BIND_TOSTRING(TransformPrimitive)
  );

#define MAKE_PAIR(Value) _MAKE_PAIR(TransformPrimitive::Type, Value)
  lua DEFINE_NESTED_ENUM(TransformPrimitive, Type, {
    MAKE_PAIR(MATRIX2D), MAKE_PAIR(MATRIX3D),
    MAKE_PAIR(TRANSLATEX), MAKE_PAIR(TRANSLATEY), MAKE_PAIR(TRANSLATEZ),
    MAKE_PAIR(TRANSLATE2D), MAKE_PAIR(TRANSLATE3D), MAKE_PAIR(TRANSLATE3D),
    MAKE_PAIR(SCALEX), MAKE_PAIR(SCALEY), MAKE_PAIR(SCALEZ),
    MAKE_PAIR(SCALE2D), MAKE_PAIR(SCALE3D),
    MAKE_PAIR(ROTATEX), MAKE_PAIR(ROTATEY), MAKE_PAIR(ROTATEZ),
    MAKE_PAIR(ROTATE2D), MAKE_PAIR(ROTATE3D),
    MAKE_PAIR(SKEWX), MAKE_PAIR(SKEWY), MAKE_PAIR(SCALE2D),
    MAKE_PAIR(PERSPECTIVE), MAKE_PAIR(DECOMPOSEDMATRIX4),
  });
#undef MAKE_PAIR

  lua.DEFINE_USERTYPE(Transform, 
    sol::call_constructor,
    sol::constructors<
      Transform(),
      Transform(Transform::PrimitiveList)
    >(),

    "makeProperty", Transform::MakeProperty,

    "clearPrimitives", &Transform::ClearPrimitives,
    "addPrimitive", &Transform::AddPrimitive,
    "getNumPrimitives", &Transform::GetNumPrimitives,
    "getPrimitive", &Transform::GetPrimitive,
    "getPrimitives", sol::overload(
      sol::resolve<Transform::PrimitiveList &()>(&Transform::GetPrimitives),
      sol::resolve<const Transform::PrimitiveList &() const>(&Transform::GetPrimitives)
    ),

    BIND_TOSTRING(Transform)
  );
  // clang-format on
}
